File: app/src/androidTest/java/com/abishov/hexocat/common/espresso/BottomNavigationViewMatcher.java
package com.abishov.hexocat.common.espresso;

import static org.hamcrest.core.Is.is;

import android.support.design.internal.BottomNavigationItemView;
import android.support.test.espresso.matcher.BoundedMatcher;
import android.view.View;
import org.hamcrest.Description;
import org.hamcrest.Matcher;

public final class BottomNavigationViewMatcher {

  public static Matcher<View> isNavigationItemChecked() {
    return withNavigationItemState(is(true));
  }

  public static Matcher<View> withNavigationItemState(final Matcher<Boolean> isChecked) {
    return new BoundedMatcher<View, BottomNavigationItemView>(BottomNavigationItemView.class) {

      @Override
      public void describeTo(Description description) {
        description.appendText("with navigation item state: ");
        isChecked.describeTo(description);
      }

      @Override
      protected boolean matchesSafely(BottomNavigationItemView item) {
        return isChecked.matches(item.getItemData().isChecked());
      }
    };
  }
}

File: android/src/androidTest/java/com/abishov/hexocat/android/common/espresso/BottomNavigationViewMatcher.kt
package com.abishov.hexocat.android.common.espresso

import android.view.View
import androidx.test.espresso.matcher.BoundedMatcher
import com.google.android.material.bottomnavigation.BottomNavigationItemView
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.core.Is.`is`

object BottomNavigationViewMatcher {

    val isNavigationItemChecked: Matcher<View>
        get() = withNavigationItemState(`is`(true))

    private fun withNavigationItemState(isChecked: Matcher<Boolean>): Matcher<View> {
        return object :
            BoundedMatcher<View, BottomNavigationItemView>(BottomNavigationItemView::class.java) {

            override fun describeTo(description: Description) {
                description.appendText("with navigation item state: ")
                isChecked.describeTo(description)
            }

            override fun matchesSafely(item: BottomNavigationItemView): Boolean {
                return isChecked.matches(item.itemData.isChecked)
            }
        }
    }
}
